##============================================================================##
##                         LEVELS OF EVIDENCE                                 ##
##----------------------------------------------------------------------------##
##                                                                            ##
##	         Set of functions to classify gene-drug interactions              ##
## 	           from GDKD and CIViC into levels of evidence.                   ##
##                                                                            ##
##============================================================================##

# druggable_onco = c
# cancer_onco = 'lymphoma'

# druggable_gdkd = druggable
# cancer_gdkd= 'lymphoma'

# druggable_civic = b
# cancer_civic = 'lymphoma'

get_levels_GDKD  = function(druggable_gdkd, cancer_gdkd){
  ## Classifies gene-drug interactions found by match_XXX_GDKD into levels of evidence. 
  ## input: druggable_gdkd is a dataframe generated by function match_XXX_GDKD.
  ## input: cancer_gdkd is a character vector with a cancer type used by GDKD.
  ## output: returns the input druggable_gdkd as a list with 10 positions. 
  ##         Each position corresponds to a level of evidence, following the order: 
  ##         A1, A2a, A2b, A2c, A3, B1, B2a, B2b, B2c, B3. 
  
  ## Check for empty input
  if (length(druggable_gdkd) == 0 || nrow(as.matrix(druggable_gdkd)) == 0){
    message("Input passed to get_levels_GDKD is empty.")
    X          = data.frame(matrix(ncol=12,nrow=0), stringsAsFactors = F)
    colnames(X) = c("Disease","Gene","Patient_variant","Variant","Description","Effect","Association","Therapeutic.context","Status","Evidence","PMID_1", "Origin Database")

    return(list(X,X,X,X,X,X,X,X,X,X))
  }
  
  druggable_gdkd[] = lapply(druggable_gdkd, as.character)
  row.names(druggable_gdkd)  = NULL
  druggable_gdkd = druggable_gdkd[,c(1,2,46,3:45)]

  ## Rows with same cancer
  same_cancer=c()
  for (ck in cancer_gdkd){
    same_cancer = c(which(druggable_gdkd$Disease==ck),same_cancer)
  }
  ## Rows with different cancer
  diff_cancer=setdiff(1:nrow(druggable_gdkd), same_cancer)

  ## Generate variables
  header            = c("Disease","Gene","Patient_variant","Variant","Description","Effect","Association","Therapeutic.context","Status","Evidence","PMID_1", "Origin Database")
  levelA1           = data.frame(matrix(ncol=12,nrow=0), stringsAsFactors = F)
  colnames(levelA1) = header
  levelA2           = data.frame(matrix(ncol=12,nrow=0), stringsAsFactors = F)
  colnames(levelA2) = header
  levelA3           = data.frame(matrix(ncol=12,nrow=0), stringsAsFactors = F)
  colnames(levelA3) = header
  levelB1           = data.frame(matrix(ncol=12,nrow=0), stringsAsFactors = F)
  colnames(levelB1) = header
  levelB2           = data.frame(matrix(ncol=12,nrow=0), stringsAsFactors = F)
  colnames(levelB2) = header
  levelB3           = data.frame(matrix(ncol=12,nrow=0), stringsAsFactors = F)
  colnames(levelB3) = header
  
  ## A levels: Same cancer rows
  for (c in same_cancer){
    ## A1 (consensus)
    consensus = which(grepl("consensus", druggable_gdkd[c,]) & !grepl("emerging", druggable_gdkd[c,]))
    if (length(consensus)!=0){
      for (l in consensus){
        levelA1[nrow(levelA1)+1,] = druggable_gdkd[c,c(1:6,(l-3):(l+1))]
        row.names(levelA1)=NULL
      }
      levelA1[,"Origin Database"] = "GDKD"
    }
   
    ## A2 and A3 (emerging)
    emerging = which(grepl("emerging", druggable_gdkd[c,]))
    if (length(emerging)!=0){
      for (l in emerging){
        levelA2[nrow(levelA2)+1,] = druggable_gdkd[c,c(1:6,(l-3):(l+1))]
        row.names(levelA2)=NULL
      }
      levelA2[,"Origin Database"] = "GDKD"
    }
  }


  ## B levels: Different cancer rows
  for (d in diff_cancer){
    ## B1 (consensus)
    consensus = which(grepl("consensus", druggable_gdkd[d,]) & !grepl("emerging", druggable_gdkd[d,]))
    if (length(consensus)!=0){
      for (l in consensus){
        levelB1[nrow(levelB1)+1,] = druggable_gdkd[d,c(1:6,(l-3):(l+1))]
        row.names(levelB1)=NULL
      }
      levelB1[,"Origin Database"] = "GDKD"
    }
    ## B2 and B3 (emerging)
    emerging = which(grepl("emerging", druggable_gdkd[d,]))
    if (length(emerging)!=0){
      for (l in emerging){
        levelB2[nrow(levelB2)+1,] = druggable_gdkd[d,c(1:6,(l-3):(l+1))]
        row.names(levelB2)=NULL
      }
      levelB2[,"Origin Database"] = "GDKD"
    }
  }
  
  ## Subset emerging evidence into specific levels
  levelA2a = levelA2[which(levelA2$Status=="late trials"),]
  levelA2b = levelA2[which(levelA2$Status=="early trials"),]
  levelA2c = levelA2[which(levelA2$Status=="case report"),]
  levelA3  = levelA2[which(levelA2$Status=="preclinical"),]
  levelB2a = levelB2[which(levelB2$Status=="late trials"),]
  levelB2b = levelB2[which(levelB2$Status=="early trials"),]
  levelB2c = levelB2[which(levelB2$Status=="case report"),]
  levelB3  = levelB2[which(levelB2$Status=="preclinical"),]

  return_list = list(levelA1,levelA2a,levelA2b,levelA2c,levelA3,levelB1,levelB2a,levelB2b,levelB2c,levelB3)

  return(return_list)
}

get_levels_CIVIC  = function(druggable_civic, cancer_civic){
  ## Classifies gene-drug interactions found by match_XXX_CIVIC into levels of evidence. 
  ## input: druggable_civic is a dataframe generated by function match_XXX_CIVIC.
  ## input: cancer_civic is a character vector with a cancer type used by CIVIC.
  ## output: returns the input druggable_civic as a list with 10 positions. 
  ##         Each position corresponds to a level of evidence, following the order: 
  ##         A1, A2a, A2b, A2c, A3, B1, B2a, B2b, B2c, B3. 
  
  ## Check for empty input
  if (length(druggable_civic) == 0 || nrow(as.matrix(druggable_civic)) == 0){
    message("Input passed to get_levels_CIVIC is empty.")
    X = data.frame(matrix(ncol=12,nrow=0), stringsAsFactors = F)
    colnames(X) = c("Disease","Gene","Patient_variant","Variant","Description","Effect","Association","Therapeutic.context","Status","Evidence","PMID_1", "Origin Database")

    return(list(X,X,X,X,X,X,X,X,X,X))
  }

  ## Remove factors
  druggable_civic[]= lapply(druggable_civic, as.character)
  row.names(druggable_civic)=NULL
  
  ## Rows with same cancer
  same_cancer=c()
  for (ck in cancer_civic){
    same_cancer = c(which(druggable_civic$disease==ck),same_cancer)
  }
  ## Rows with different cancer
  other_cancer  = setdiff(1:nrow(druggable_civic),same_cancer)
  
  ## Add CIVIC tag to druggable_civic
  druggable_civic["database_origin"] = "CIVIC"

  ## Generate variables
  levelA1            = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F)
  levelA2a           = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F)
  levelA2c           = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F)
  levelA3            = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F)
  levelB1            = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F)
  levelB2a           = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F)
  levelB2c           = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F)
  levelB3            = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F)
  
  ## Subset same cancer rows into A levels
  levelA1  = druggable_civic[intersect(same_cancer,which(druggable_civic$evidence_level=="A")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id", "database_origin")]
  levelA2a = druggable_civic[intersect(same_cancer,which(druggable_civic$evidence_level=="B")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id", "database_origin")]
  levelA2c = druggable_civic[intersect(same_cancer,which(druggable_civic$evidence_level=="C")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id", "database_origin")]
  levelA3  = druggable_civic[intersect(same_cancer,which(druggable_civic$evidence_level=="D")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id", "database_origin")]

  ## Subset different cancer rows into B levels
  levelB1  = druggable_civic[intersect(other_cancer,which(druggable_civic$evidence_level=="A")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id", "database_origin")]
  levelB2a = druggable_civic[intersect(other_cancer,which(druggable_civic$evidence_level=="B")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id", "database_origin")]
  levelB2c = druggable_civic[intersect(other_cancer,which(druggable_civic$evidence_level=="C")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id", "database_origin")]
  levelB3  = druggable_civic[intersect(other_cancer,which(druggable_civic$evidence_level=="D")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id", "database_origin")]

  ## Homogeneize colnames with GDKD 
  colnames(levelA1)  = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1", "Origin Database")
  colnames(levelA2a) = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1", "Origin Database")
  colnames(levelA2c) = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1", "Origin Database")
  colnames(levelA3)  = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1", "Origin Database")
  colnames(levelB1)  = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1", "Origin Database")
  colnames(levelB2a) = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1", "Origin Database")
  colnames(levelB2c) = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1", "Origin Database")
  colnames(levelB3)  = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1", "Origin Database")

  return_list = list(levelA1,levelA2a,data.frame(matrix(ncol=9,nrow=0)),levelA2c,levelA3,levelB1,levelB2a,data.frame(matrix(ncol=9,nrow=0)),levelB2c,levelB3)

  return(return_list)
}

get_levels_ONCO  = function(druggable_onco, cancer_onco){
  ## Classifies gene-drug interactions found by match_XXX_ONCO into levels of evidence. 
  ## input: druggable_onco is a dataframe generated by function match_XXX_ONCO.
  ## input: cancer_onco is a character vector with a cancer type used by OncoKB.
  ## output: returns the input druggable_onco as a list with 10 positions. 
  ##         Each position corresponds to a level of evidence, following the order: 
  ##         A1, A2a, A2b, A2c, A3, B1, B2a, B2b, B2c, B3. 
  
  ## Check for empty input
  if (length(druggable_onco) == 0 || nrow(as.matrix(druggable_onco)) == 0){
    # message("Input passed to get_levels_ONCO is empty.")
    X = data.frame(matrix(ncol=12,nrow=0), stringsAsFactors = F)
    colnames(X) = c("Disease","Gene","Variant","Patient_variant","Description","Effect","Association","Therapeutic.context","Status","Evidence","PMID_1", "Origin Database")
    return(list(X,X,X,X,X,X,X,X,X,X)) #Why? Idk what the point of this is
  }
  
  ## Remove factors
  druggable_onco[]= lapply(druggable_onco, as.character)
  row.names(druggable_onco)=NULL
  
  ## Only keep interesting columns from druggable_onco - Add association column
  colnames(druggable_onco) = c("Gene", "Entrez_ID", "Variant", "Disease", "Drugs", "Evidence_level", "PMID", "Abstract", "Isoform", "RefSeq", "Patient_variant", "Origin_Database")
  druggable_onco$Association = ""
  for (i in 1:nrow(druggable_onco)) {
    if (grepl("R1|R2",druggable_onco[i,"Evidence_level"])) {
      druggable_onco[i,"Association"] = "Resistance"
    }
    else {
      druggable_onco[i,"Association"] = "Sensitivity/Response"
    }
  }
  druggable_onco = druggable_onco[c(1,2,11,3,4,5,6,7,12,13)]
  
  ## Rows with same cancer
  same_cancer=c()
  for (ck in cancer_onco){
    same_cancer = c(which(druggable_onco$Disease==ck),same_cancer)
  }
  ## Rows with different cancer
  other_cancer  = setdiff(1:nrow(druggable_onco),same_cancer)
  
  ## Generate variables
  levelA1            = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F)
  levelA2            = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F)
  levelA3            = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F) ## Added by kko
  levelB1            = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F) ## Added by kko
  levelB2            = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F)
  levelB3            = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F)
  levelU1            = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F) # For the OncoKB rows that have R1 or R2
  
  ## Subset same cancer rows into A levels
  ## kko: Changed OR Opterator ('||') to a vectorized OR operator ('|') based on https://stackoverflow.com/questions/15140944/r-which-statement-with-multiple-conditions
  levelA1  = druggable_onco[intersect(same_cancer,which(druggable_onco$Evidence_level=="1")),c("Disease","Gene","Patient_variant", "Variant","Association","Drugs","Evidence_level","PMID", "Origin_Database")]
  if(nrow(levelA1) > 0){
        levelA1$Evidence_level = c("A1")
  }

  
  levelA2  = druggable_onco[intersect(same_cancer,which(druggable_onco$Evidence_level=="2A" | druggable_onco$Evidence_level=="3A")),c("Disease","Gene","Patient_variant", "Variant","Association","Drugs","Evidence_level","PMID", "Origin_Database")]
  if(nrow(levelA2) > 0){
      levelA2$Evidence_level = "A2"
  }
  
  levelA3 = druggable_onco[intersect(same_cancer,which(druggable_onco$Evidence_level=="3B" | druggable_onco$Evidence_level=="4")),c("Disease","Gene","Patient_variant", "Variant","Association","Drugs","Evidence_level","PMID", "Origin_Database")]
  if(nrow(levelA3) > 0){
      levelA3$Evidence_level = "A3"
  }

  ## Subset different cancer rows into B levels
  levelB1  = druggable_onco[intersect(other_cancer,which(druggable_onco$Evidence_level=="1")),c("Disease","Gene","Patient_variant", "Variant","Association","Drugs","Evidence_level","PMID", "Origin_Database")]
  if(nrow(levelB1) > 0){
      levelB1$Evidence_level = "B1"
  }
  
  levelB2  = druggable_onco[intersect(other_cancer,which(druggable_onco$Evidence_level=="2A" | druggable_onco$Evidence_level=="3A")),c("Disease","Gene","Patient_variant", "Variant","Association","Drugs","Evidence_level","PMID", "Origin_Database")]
  if(nrow(levelB2) > 0){
      levelB2$Evidence_level = "B2"
  }
  
  levelB3  = druggable_onco[intersect(other_cancer,which(druggable_onco$Evidence_level=="3B" |druggable_onco$Evidence_level=="4")),c("Disease","Gene","Patient_variant", "Variant","Association","Drugs","Evidence_level","PMID", "Origin_Database")]
  if(nrow(levelB3) > 0){
      levelB3$Evidence_level = "B3"
  }

  ## Subset rows with unspecified evidence levels into U levels
  levelU1  = druggable_onco[intersect(same_cancer,which(druggable_onco$Evidence_level=="R1" | druggable_onco$Evidence_level=="R2")),c("Disease","Gene","Patient_variant", "Variant","Association","Drugs","Evidence_level","PMID", "Origin_Database")]
  levelU1  = rbind(levelU1,druggable_onco[intersect(other_cancer,which(druggable_onco$Evidence_level=="R1" | druggable_onco$Evidence_level=="R2")),c("Disease","Gene","Patient_variant", "Variant","Association","Drugs","Evidence_level","PMID", "Origin_Database")])
  if(nrow(levelU1) > 0){
      levelU1$Evidence_level = "R1(OncoKB)"
  }

  ## Replace specified evidence levels
  output = rbind(levelA1,levelA2,levelA3,levelB1,levelB2,levelB3,levelU1)
  
  ## Homogenize colnames so they fit to clean_levels()
  colnames(output) = c("Cancer","Gene","Pat_Var","Known_Var","Predicts","Drugs","Evidence","Ref", "Origin_Database")
  
  return(output)
}



merge_levels = function(levels_gdkd, levels_civic){
  ## Merges the output of get_levels_GDKD and get_levels_CIVIC
  ## input: levels_gdkd is a list generated by function get_levels_GDKD.
  ## input: levels_civic is a list  generated by function get_levels_CIVIC.
  ## output: returns a list with 10 positions. 
  ##         Each position corresponds to a level of evidence, following the order: 
  ##         A1, A2a, A2b, A2c, A3, B1, B2a, B2b, B2c, B3. 

  ## Homogeneize columns of levels_gdkd as in levels_civic
  for (i in 1:10){
    levels_gdkd[[i]]     = levels_gdkd[[i]][,c(1:4,6,7:9,11,12)] 		    # Put in same order as civic columns
    levels_gdkd[[i]][,4] = paste(levels_gdkd[[i]][,4],levels_gdkd[[i]][,5]) # Paste effect to variant (eg. T45X loss-of-function)
    levels_gdkd[[i]]     = levels_gdkd[[i]][,c(1:4,6:9,10)]
  }

  ## Merge two lists, position by position
  merged = c()
  for (i in 1:10){
    merged[[i]]           = rbind(levels_gdkd[[i]],levels_civic[[i]])
    rownames(merged[[i]]) = NULL
    merged[[i]]           = merged[[i]][order(merged[[i]][,"Gene"]),] #in each level, order by gene
    rownames(merged[[i]]) = NULL
  }
  
  return(merged)

}

# A = return_list

clean_levels = function(A,synonyms,sort_by="levels"){ #removed O argument after A, removed underscores in col naming
  ## Creates a clean table with gene-drug interactions from list of levels.
  ## input: A is list generated by functions get_levels_GDKD, get_levels_CIVIC or merge_levels.
  ## The possible options for the argument sort_by are "levels", "genes" and "drug" or "drug_freq".
  ## output: returns a dataframe with a clean annotation of gene-drug interactions. 
  
  if (is.list(A)){
    A1  = A[[1]]
    A2a = A[[2]]
    A2b = A[[3]]
    A2c = A[[4]]
    A3  = A[[5]]
    B1  = A[[6]]
    B2a = A[[7]]
    B2b = A[[8]]
    B2c = A[[9]]
    B3  = A[[10]]
    A = rbind(A1,B1,A2a,A2b,A2c,B2a,B2b,B2c,A3,B3)
  }
  
  ## To avoid error if A is empty
  if (length(A) == 0){
      B = data.frame(matrix(ncol=10,nrow=0), stringsAsFactors = F)
      colnames(B) = c("Cancer","Gene","Pat_Var","Known_Var","Predicts","Drugs","Evidence","Ref", "Origin_Database", "level")
      return(B)
  }
  
  onco_exists = FALSE
  if (exists("O") && !is.null(O)) {
      if (nrow(O) >= 1) {
         onco_exists = TRUE
      }
  } 
  
  ## kko: this stuff is already done in "get_levels_ONCO()"
  ## Convert OncoKB Evidence Levels to ours
  # if (onco_exists) {
  #     rownames(O) = NULL
  #     for (i in 1:nrow(O)) {
  #         if (O[i,"Evidence"] == "1") {
  #             O[i,"Evidence"] = "A1"
  #         }
  #         else if (O[i,"Evidence"] == "2A" || O[i,"Evidence"] == "3A") {
  #             O[i,"Evidence"] = "A2"
  #         }
  #         else if (O[i,"Evidence"] == "3B") {
  #             O[i,"Evidence"] = "B2"
  #         }
  #         else if (O[i,"Evidence"] == "4") {
  #             O[i,"Evidence"] = "B3"
  #         }
  #         else if (O[i,"Evidence"] == "R1") {
  #             O[i,"Evidence"] = "R1(OncoKB)"
  #         }
  #         else {
  #             O[i,"Evidence"] = "R2(OncoKB)"
  #         }
  #     }
  # }
  
  
  ## Homogeneize evidence of civicDB to knowledgeDB categories
  A$Status = gsub(pattern = "A$",replacement = "approved",A$Status)
  A$Status = gsub(pattern = "B$",replacement = "clin_trials",A$Status)
  A$Status = gsub(pattern = "C$",replacement = "case_report",A$Status)
  A$Status = gsub(pattern = "D$",replacement = "preclinical",A$Status)
  
  
  ## Remove non informative columns
  A$Effect      = NULL
  A$Evidence    = NULL
  A$Description = NULL
  rownames(A)   = NULL
  
  ## Add column names and levels of evidence
  colnames(A) = c("Cancer","Gene","Pat_Var","Known_Var","Predicts","Drugs","Evidence","Ref", "Origin_Database")
  A$level     = c(rep("A1",nrow(A1)),rep("B1",nrow(B1)),rep("A2",nrow(A2a)),rep("A2",nrow(A2b)),rep("A2",nrow(A2c)),rep("B2",nrow(B2a)),rep("B2",nrow(B2b)),rep("B2",nrow(B2c)),rep("A3",nrow(A3)),rep("B3",nrow(B3)))
  
  #print("Check A in clean_levels fn")
  #print(head(A))

  ## Add OncoKB to the A matrix
 # O$level     = O$Evidence
 # O$Evidence = gsub(pattern = "A1$", replacement = "approved", O$Evidence)
 # O$Evidence = gsub(pattern = "A2$", replacement = "clin. trials", O$Evidence)
 # O$Evidence = gsub(pattern = "A3$", replacement = "preclinical", O$Evidence)
 # O$Evidence = gsub(pattern = "B1$", replacement = "approved", O$Evidence)
 # O$Evidence = gsub(pattern = "B2$", replacement = "clin. trials", O$Evidence)
 # O$Evidence = gsub(pattern = "B3$", replacement = "preclinical", O$Evidence)
 # O$Evidence = gsub(pattern = "R.*$", replacement = "unspecified", O$Evidence)
 # A = rbind(A,O)

  ## Prepare disease name homogenization
  civ_disease = lapply(A$Cancer,function(x) {
    index = grep(x,synonyms$civic,ignore.case = TRUE)
    return(as.character(synonyms[index,"Report"]))
  })
  
  gdkd_disease = lapply(A$Cancer,function(x) {
    index = grep(x,synonyms$knowledge,ignore.case = TRUE)
    return(as.character(synonyms[index,"Report"]))
  })
  
  #onco_disease = lapply(A$Cancer,function(x) {
  #  index = grep(x,synonyms$oncokb,ignore.case = TRUE)
  #  return(as.character(synonyms[index,"Report"]))
  #})

  
  ## Homogenize disease names
  # civ_disease = sapply(civ_disease,function(x){as.character(x[1])})
  # gdkd_disease = sapply(gdkd_disease,function(x){as.character(x[1])})
  # onco_disease = sapply(onco_disease,function(x){as.character(x[1])})
  # civ_disease[is.na(civ_disease)] = gdkd_disease[which(is.na(civ_disease))]
  # civ_disease[is.na(civ_disease)] = onco_disease[which(is.na(civ_disease))]
  # A$Cancer   = civ_disease
  
  ## Shorten "amplification" and "deletion"
  A$Pat_Var = gsub(pattern = "amplification",replacement = "ampl.",A$Pat_Var)
  A$Pat_Var = gsub(pattern = "deletion",replacement = "del.",A$Pat_Var)
  A$Pat_Var = gsub(pattern = "_",replacement = "-",A$Pat_Var)
  
  ## Variants & Association to lower case
  upper               = grep("[A-Za-z][0-9]+",A$Known_Var)
  if (length(upper) !=0 ){
    lower               = tolower(A[-upper,"Known_Var"])
    A[-upper,"Known_Var"] = lower
  }
  if (length(upper) ==0 ){
    A$Known_Var = tolower(A$Known_Var)
  }
  
  ## Remove "_" and ^ from PMID
  A$Ref = gsub(pattern = "_",replacement = " ",A$Ref)
  A$Ref = gsub(pattern = "^",replacement = " ",A$Ref, fixed=TRUE)
  
  ## Shorten and unify `Known Var`s
  A$Known_Var       = gsub(pattern = "_",replacement = "-",A$Known_Var)
  A$Known_Var       = gsub(pattern = "any",replacement = "any variant",A$Known_Var, ignore.case=TRUE)
  A$Known_Var       = gsub(pattern = "amplification",replacement = "ampl.",A$Known_Var, ignore.case=TRUE)
  A$Known_Var       = gsub(pattern = "deletion",replacement = "del.",A$Known_Var, ignore.case=TRUE)
  A$Known_Var       = gsub(pattern = "mutation",replacement = "mut.",A$Known_Var, ignore.case=TRUE)
  A$Known_Var       = gsub(pattern = "mut[\\.]*",replacement = "any mut.",A$Known_Var, ignore.case=TRUE)
  A$Known_Var       = gsub(pattern = "gain-of-function",replacement = "(GoF)",A$Known_Var, ignore.case=TRUE)
  A$Known_Var       = gsub(pattern = "loss-of-function",replacement = "(LoF)",A$Known_Var, ignore.case=TRUE)
  A$Known_Var       = gsub(pattern = "not applicable",replacement = "",A$Known_Var, ignore.case=TRUE)
  #A$`Known Var`       = gsub(pattern = ",",replacement = ", ",A$`Known Var`, ignore.case=TRUE)
  A$Predicts   = tolower(A$Predicts)

  ## Aggregate duplicate rows
  A_agg   = aggregate(A, by = list(A$Cancer,A$Gene,A$Pat_Var,A$Predicts,A$Drugs,A$Evidence), #Errors out here, undefined columns selected 04/04/23
                      FUN = function(X) paste(unique(X), collapse=", "))[,7:16]                             
  A=A_agg
  
  #Reorganize table
  A = A[,c(2,3,1,4:10)] #Gene, Pat Var, Cancer, Known Var, Predicts, Drugs, Evidence, Ref, OriginDB, level
  
  
  #Reorder rows according drugs
  if(sort_by=="drug_freq"){
    d = unlist(strsplit(A$Drugs," |,|\\)|\\("))
    d = gsub("^ ","",d)
    d = gsub(" $","",d)
    d = tolower(d)
    for (i in c ("","inhibitors","inhibitor","pathway","+","mabs","tkis","in","mutant","blockade","tumor","tumors","plus","wt","setting","or","c","therapy","hormone","monoclonal")){
      d[which(d==i)] = NA #jetzt hat sie keine kürzere liste, sondern eine liste mit NAs. ist das gewollt?
    }
    d = na.omit(d) #liste ist gekürzt
    drug_order=c()
    for (drug in names(sort(table(d),decreasing = TRUE))){
      index = grep(drug,tolower(A[,6]),ignore.case = T)
      index = index[!(index %in% drug_order)]
      drug_order = c(drug_order,index)
    }
    A = A[drug_order,]
  }
  
  #Reorder rows according genes
  if(sort_by=="genes"){
    A = A[order(A$Gene, A$level),]
    
  }
  
  #Reorder rows by evidence levels
  if(sort_by=="levels"){
    A = A[order(A$level, A$Gene),]
  }
  
  #A  =  aggregate(A,
  #  by=list(A$Gene,A$Cancer,A$Drugs,A$level,A$Predicts),FUN = function(X) paste(X, collapse=","))[,6:14]
  return(A)
}  



summary_actionable_genes = function(A,maf=data.frame(),cnv=data.frame(),druggableTARGET=data.frame(), druggableONCO = data.frame()){
  ## Creates a summary table of actionable genes and genomic quality, of provided by user.
  ## input: A is dataframe generated by clean_levels()
  ## input: maf is a data frame
  ## input: cnv is a 
  ## input: druggableTARGET is a dataframe generated by function match_TARGET_MERIC()
  ## input: druggableONCO is a dataframe generated by function match_XXX_OncoKb
  ## output: returns two dataframes (SNVs and CNVs) with a summary of actionable genes (gene, variant, level of evidence, genomic quality)
 tab_summary_snvs=data.frame()
 tab_summary_cnvs=data.frame()

if (nrow(A)!=0|| nrow(druggableTARGET)!=0 || nrow(druggableONCO != 0)){
  tab_summary           <- A[,c("Gene","Pat_Var","level")]
  tab_summary$level.col <- tab_summary$level
  if (nrow(druggableTARGET)>0){
     B            <- druggableTARGET[,c("Gene","Patient_variant")]
     B$Patient_variant <- gsub(pattern = "amplification",replacement = "ampl.", B$Patient_variant)
     B$Patient_variant <- gsub(pattern = "deletion",replacement = "del.",B$Patient_variant)
     B$level      <- "unknown"
     B$level.col  <- "unknown"
     colnames(B)  <- c("Gene","Pat_Var","level","level.col")
     B            <- B[!(B$Gene %in% tab_summary$Gene),]
     tab_summary  <- rbind(tab_summary,B)
  }
 # if (nrow(druggableONCO) >0) {
 #   C             <- druggableONCO[,c("Gene", "Patient_variant")]
 #   C$Patient_variant <- gsub(pattern = "amplification",replacement = "ampl.", C$Patient_variant)
 #   C$Patient_variant <- gsub(pattern = "deletion",replacement = "del.",C$Patient_variant)
 #   for (i in 1:nrow(druggableONCO)) {
 #     C[i,"level"] = paste("OncoKB", druggableONCO[i,"Evidence.Level"])
 #   }
 #   C$level.col  <- "unknown"
 #   colnames(C)  <- c("Gene","Pat Var","level","level.col")
 #   C            <- C[!(C$Gene %in% tab_summary$Gene),]
 #   tab_summary  <- rbind(tab_summary,C)
 # }
  tab_summary$Pat_Var <- sapply(tab_summary$Pat_Var , function(x) gsub("NEW", "", x))
  tab_summary$Pat_Var <- sapply(tab_summary$Pat_Var , function(x) gsub("([A-Z]) ([A-Z])", "\\1, \\2", x))
  tab_summary$Pat_Var <- sapply(tab_summary$Pat_Var , function(x) gsub(" ", "", x))
  tab_summary           <- aggregate(tab_summary[,c("Gene","Pat_Var","level.col")],by=list(name=tab_summary$Gene),FUN = function(X) paste(unique(X), collapse=","))[,-1]

  ## SNVs

  if(!is.null(maf)){
    if (nrow(maf)!=0){
      snvs <- tab_summary$Gene %in% maf[,1]
      i <- na.omit(match(tab_summary$Gene,maf[,1]))
      tab_summary_snvs <- cbind(tab_summary[snvs,],maf[i,setdiff(1:ncol(maf),1:3)])
      twice <- grep(",",tab_summary_snvs$Pat_Var)
      if (length(twice)!=0){
        for (n in 1:length(twice)){
          vars=c()
          vars=strsplit(tab_summary_snvs[twice[n],"Pat_Var"],",")[[1]]
          vars=unique(vars)
          if(grepl("ampl\\.|del\\.",paste(vars,collapse=" "))){
            tab_summary_snvs[twice[n],"Pat_Var"]=vars[grep("ampl\\.|del\\.",vars,invert=T)]
          }
          else{
            tab_summary_snvs[twice[n],"Pat_Var"]=paste(vars,collapse=", ")
          }
        } 
      }
    }
    colnames(tab_summary_snvs) <- c("Gene","Patient's Variant","Level of Evidence",colnames(maf)[setdiff(1:ncol(maf),1:3)])
    #colnames(tab_summary_snvs) <- paste("\\textbf{",colnames(tab_summary_snvs),"}")

  }


 ## CNVs

  tab_summary_cnvs <- tab_summary[grep("ampl\\.|del\\.",tab_summary$Pat_Var),]
  if (!is.null(tab_summary_cnvs) & nrow(tab_summary_cnvs)!=0 ){
    j <- na.omit(match(tab_summary_cnvs$Gene,cnv[,1]))
    tab_summary_cnvs <- cbind(tab_summary_cnvs,cnv[j,setdiff(1:ncol(cnv),1:2)])
    twice <- grep(",",tab_summary_cnvs$Pat_Var)
    for (t in twice){
      vars=strsplit(tab_summary_cnvs[t,"Pat_Var"],", ")[[1]]
      tab_summary_cnvs[t,"Pat_Var"]=vars[grep("ampl\\.|del\\.",vars)]
    }

    colnames(cnv) = gsub("\\^|_"," ",colnames(cnv))
    colnames(tab_summary_cnvs) <- c("Gene","Patient's Variant","Level of Evidence",colnames(cnv)[setdiff(1:ncol(cnv),1:2)])
    #colnames(tab_summary_cnvs) <- paste("\\textbf{",colnames(tab_summary_cnvs),"}")
  }

}
return(list(tab_summary_snvs,tab_summary_cnvs))
}
